File: pincer/tools/ls.go
package tools

import (
	"fmt"
	"os"
	"strings"
)

type LsParams struct {
	Path string `json:"path"`
}

func Ls(p *LsParams) (*Result, error) {
	entries, err := os.ReadDir(p.Path)
	if err != nil {
		return &Result{Content: "Error listing directory: " + err.Error(), IsError: true}, nil
	}

	if len(entries) == 0 {
		return &Result{Content: "(empty directory)"}, nil
	}

	var lines []string
	for _, entry := range entries {
		prefix := "[FILE]"
		size := ""

		if entry.IsDir() {
			prefix = "[DIR]"
		} else {
			info, err := entry.Info()
			if err == nil {
				size = fmt.Sprintf(" (%s)", formatSize(info.Size()))
			}
		}

		lines = append(lines, fmt.Sprintf("%s %s%s", prefix, entry.Name(), size))
	}

	return &Result{Content: strings.Join(lines, "\n")}, nil
}

func formatSize(bytes int64) string {
	switch {
	case bytes < 1024:
		return fmt.Sprintf("%dB", bytes)
	case bytes < 1024*1024:
		return fmt.Sprintf("%.1fKB", float64(bytes)/1024)
	case bytes < 1024*1024*1024:
		return fmt.Sprintf("%.1fMB", float64(bytes)/(1024*1024))
	default:
		return fmt.Sprintf("%.1fGB", float64(bytes)/(1024*1024*1024))
	}
}

File: src/tools/ls.ts
import fs from "node:fs/promises";
import path from "node:path";
import type { ToolResult } from "./bash.js";

export interface LsParams {
  path: string;
}

export async function ls(params: LsParams): Promise<ToolResult> {
  try {
    const entries = await fs.readdir(params.path, { withFileTypes: true });

    const lines: string[] = [];
    for (const entry of entries) {
      const prefix = entry.isDirectory() ? "[DIR]" : "[FILE]";
      let size = "";

      if (!entry.isDirectory()) {
        try {
          const stat = await fs.stat(path.join(params.path, entry.name));
          size = ` (${formatSize(stat.size)})`;
        } catch {
          // ignore stat errors
        }
      }

      lines.push(`${prefix} ${entry.name}${size}`);
    }

    if (lines.length === 0) {
      return { content: "(empty directory)" };
    }

    return { content: lines.join("\n") };
  } catch (err: any) {
    return {
      content: `Error listing directory: ${err.message}`,
      isError: true,
    };
  }
}

function formatSize(bytes: number): string {
  if (bytes < 1024) return `${bytes}B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)}KB`;
  if (bytes < 1024 * 1024 * 1024)
    return `${(bytes / (1024 * 1024)).toFixed(1)}MB`;
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(1)}GB`;
}
